File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        l = 0
        sets = set ()
        ret = 0

        for r in range (len (s)):
            while s[r] in sets: 
                sets.remove (s[l])
                l+=1

            sets.add (s[r])

            ret = max (ret, r - l + 1)

        return ret

        

File: Data Structures & Algorithms/string-encode-and-decode/submission-1.go


type Solution struct{}

func (s *Solution) Encode(strs []string) string {
    var ret strings.Builder

    for _, str := range strs {
        length := strconv.Itoa(len(str))

        ret.WriteString(length)
        ret.WriteString("#")
        ret.WriteString(str)
    }
    return ret.String()
}

func (s *Solution) Decode(encoded string) []string {
	result := []string{}
	i := 0

	for i < len (encoded) {
		// find the #
		j:=i
		for encoded[j] != '#' {
			j++
		}

		// get the length number
		lengthStr := encoded [i:j]

		length,_ :=strconv.Atoi(lengthStr)

		//move past #
		j++

		word := encoded [j:j+length]
		result = append (result, word)

		i = j+length
	}

	return result

}
